File: course-mern-api/src/middleware/rateLimiter.js
// create a ratelimiter that allows 10 requests per 20 seconds
// import { Ratelimit } from '@upstash/ratelimit'
// import { Redis } from '@upstash/redis'
//
// const ratelimit = new Ratelimit({
//   redis: Redis.fromEnv(),
//   limiter: Ratelimit.slidingWindow(10, '20 s')
// })
//
// const rateLimiter = async (req, res, next) => {
//
//   try {
//     const { success } = await ratelimit.limit('my-limit-key')
//
//     if (!success) {
//       return res.status(429).json({
//         message: 'Too many requests, please try again later'
//       })
//     }
//
//     next()
//
//   } catch(error) {
//     console.log('Rate limit error', error)
//     next(error)
//   }
// }

import redis from '../config/redis.js'

// 2. Upstash의 slidingWindow 기능을 완벽히 대체하는 고성능 Lua 스크립트 정의
// (여러 노드에서 동시에 요청이 와도 race condition 없이 정확하게 동기화됩니다)
const slidingWindowLua = `
  local key = KEYS[1]
  local now = tonumber(ARGV[1])
  local window = tonumber(ARGV[2])
  local limit = tonumber(ARGV[3])
  
  -- 20초 범위를 벗어난 과거의 요청 기록 싹 청소하기
  local clearBefore = now - window
  redis.call('zremrangebyscore', key, '-inf', clearBefore)
  
  -- 현재 윈도우 안에 남아있는 요청 횟수 카운트
  local currentRequests = redis.call('zcard', key)
  
  if currentRequests < limit then
      -- 제한 미만이면 현재 시간(타임스탬프)을 기록하고 통과 (1 반환)
      redis.call('zadd', key, now, now)
      -- 자동으로 데이터가 청소되도록 만료시간 설정 (윈도우 크기 + 1초 마진)
      redis.call('expire', key, math.ceil(window / 1000) + 1)
      return 1
  else
      -- 제한 초과 시 차단 (0 반환)
      return 0
  end
`

// ioredis에 커스텀 명령어 등록 (이름: performRateLimit)
redis.defineCommand('performRateLimit', {
  numberOfKeys: 1,
  lua: slidingWindowLua,
})

const rateLimiter = async (req, res, next) => {
  try {
    // 실무에서는 보통 'rate:limit:' + req.ip 또는 유저 ID 조합을 키로 씁니다.
    const limitKey = 'rate:my-rate-limit'

    const now = Date.now()               // 현재 시간 (Millisecond)
    const windowMs = 20 * 1000           // 20초 (Upstash의 '20 s'와 동일)
    const maxRequests = 10               // 최대 허용 횟수

    // 레디스 내부에서 원자적으로 슬라이딩 윈도우 연산 실행
    const result = await redis.performRateLimit(limitKey, now, windowMs, maxRequests)
    const success = result === 1

    if (!success) {
      return res.status(429).json({
        message: 'Too many requests, please try again later'
      })
    }

    next()

  } catch(error) {
    console.log('Rate limit error', error)
    next(error)
  }
}

export default rateLimiter

File: custom-scala-api/src/main/scala/notes/core/RateLimiter.scala
package notes.core

import cats.effect.IO

import dev.profunktor.redis4cats.RedisCommands
import dev.profunktor.redis4cats.effects.ScriptOutputType

import java.time.Instant

trait RateLimiter {

  /** 주어진 키에 대한 요청이 처리 한도(rate limit) 내에 있는지 확인합니다.
    *
    * @param key
    *   고유 키 (예: IP 주소 또는 사용자 ID)
    * @param maxRequests
    *   해당 시간 내에 허용되는 최대 요청 수
    * @param windowSeconds
    *   제한 시간(윈도우)의 길이
    * @return
    *   허용된 경우 true, 한도를 초과한 경우 false
    */
  def isAllowed(key: String, maxRequests: Int, windowSeconds: Int): IO[Boolean]
}

// 이미 연결된 레디스 커넥션 명령어를 생성자로 주입받습니다.
class LiveRateLimiter private (redisCmd: RedisCommands[IO, String, String]) extends RateLimiter {

  // 원자성(Atomicity)을 보장하는 슬라이딩 윈도우 Lua 스크립트
  private val slidingWindowLua: String =
    """
      |local key = KEYS[1]
      |local now = tonumber(ARGV[1])
      |local window = tonumber(ARGV[2])
      |local limit = tonumber(ARGV[3])
      |
      |-- 윈도우 범위를 벗어난 오래된 타임스탬프 기록 삭제
      |local clearBefore = now - window
      |redis.call('zremrangebyscore', key, '-inf', clearBefore)
      |
      |-- 현재 윈도우 내의 요청 횟수 확인
      |local currentRequests = redis.call('zcard', key)
      |
      |if currentRequests < limit then
      |    -- 제한 미만이면 현재 시간 추가 후 승인 (1 리턴)
      |    redis.call('zadd', key, now, now)
      |    redis.call('expire', key, math.ceil(window / 1000) + 1)
      |    return 1
      |else
      |    -- 제한 초과 시 차단 (0 리턴)
      |    return 0
      |end
    """.stripMargin

  // 2. 요청 허용 여부를 판단하는 핵심 메서드
  override def isAllowed(key: String, maxRequests: Int, windowSeconds: Int): IO[Boolean] = {
    val limitKey = s"rate:my-rate-limit"

    // Resource 패턴을 통해 사용 후 커넥션 풀 자원을 안전하게 자동 해제

    for {
      now <- IO(Instant.now().toEpochMilli) // 현재 타임스탬프 (ms)
      windowMs = windowSeconds * 1000

      // Lua 스크립트 실행 (리턴 타입은 RedisType.Integer -> Long 매핑)
      result <- redisCmd.eval(
        script = slidingWindowLua,
        output = ScriptOutputType.Integer,
        keys = List(limitKey),
        values = List(now.toString, windowMs.toString, maxRequests.toString)
      )
    } yield result == 1L // 1이면 true(승인), 0이면 false(차단)

  }
}

object LiveRateLimiter {
  def apply(redis: RedisCommands[IO, String, String]): IO[LiveRateLimiter] =
    IO(new LiveRateLimiter(redis))
}
